src/colorizer/VectorField.ts

```diff
@@ -0,0 +1,313 @@
+    const arrowHead2: ArrayVector3 = [Math.cos(arrowAngle2), Math.sin(arrowAngle2), 0];
+
+    const lengthPx = Math.min(arrowStyle.maxHeadLengthPx * 2, vectorLengthPx);
+
+    // Keep arrow head length constant relative to onscreen pixels
+    // Arrow heads are in screen space pixel coordinates, so we divide by canvas resolution to get relative canvas coordinates,
+    // and then divide again by `frameToCanvasCoordinates` to get relative frame coordinates.
+    arrowHead1[0] =
+      (arrowHead1[0] * lengthPx) / this.frameToCanvasCoordinates.x / this.canvasResolution.x + normVectorEnd[0];
+    arrowHead1[1] =
+      (arrowHead1[1] * lengthPx) / this.frameToCanvasCoordinates.y / this.canvasResolution.y + normVectorEnd[1];
+    arrowHead2[0] =
+      (arrowHead2[0] * lengthPx) / this.frameToCanvasCoordinates.x / this.canvasResolution.x + normVectorEnd[0];
+    arrowHead2[1] =
+      (arrowHead2[1] * lengthPx) / this.frameToCanvasCoordinates.y / this.canvasResolution.y + normVectorEnd[1];
+
+    vertices.set(normVectorEnd, 0 * VERTEX_LENGTH);
+    vertices.set(arrowHead1, 1 * VERTEX_LENGTH);
+    vertices.set(normVectorEnd, 2 * VERTEX_LENGTH);
+    vertices.set(arrowHead2, 3 * VERTEX_LENGTH);
+
+    return vertices;
+  }
+
+  private getArrowVertices(centroid: [number, number], delta: [number, number]): Float32Array {
+    // Draw the main vector line segment from centroid to centroid + delta.
+    // TODO: Perform scaling as a step in a vertex shader.
+    const vertices = new Float32Array(VERTICES_PER_VECTOR_LINE * VERTEX_LENGTH);
+
+    const vectorStart: ArrayVector3 = [centroid[0], centroid[1], 0];
+    const vectorEnd: ArrayVector3 = [
+      centroid[0] + delta[0] * this.config.scaleFactor,
+      centroid[1] + delta[1] * this.config.scaleFactor,
+      0,
+    ];
+    const normVectorEnd = this.framePixelsToRelativeFrameCoords(vectorEnd);
+    const normVectorStart = this.framePixelsToRelativeFrameCoords(vectorStart);
+    vertices.set(normVectorStart, 0 * VERTEX_LENGTH);
+    vertices.set(normVectorEnd, 1 * VERTEX_LENGTH);
+
+    // Draw arrow head
+    vertices.set(this.getArrowHeadVertices(normVectorStart, normVectorEnd), 2 * VERTEX_LENGTH);
+
+    return vertices;
+  }
+
+  /**
+   * Calculates the vertices for the vector lines for ALL objects in the dataset,
+   * across ALL timepoints. Vertices are stored in the line's geometry buffer,
+   * grouped by frame number. Each pair of vertices represents a line segment.
+   *
+   * Currently, each vector line takes up 6 vertices, two for the main line segment
+   * and four for the arrow heads' two lines.
+   * For a vertex at index `i`, the vertex's XY components are stored at
+   * at `(2i, 2i + 1)`.
+   *
+   * This method updates vertices and the `timeToVertexIndexRange` map, which maps from
+   * a frame number to the range of vertices that should be drawn for that frame.
+   */
+  private updateLineVertices(): void {
```

Comment:
a personal preference I have is being able to read a function without scrolling. There is nothing wrong with this function, but personally, probably each section that has a comment over it I would have made a separate function.